test/extended/storage/driver_configuration.go

```diff
@@ -0,0 +1,326 @@
+					o.Expect(err).NotTo(o.HaveOccurred())
+
+					// Wait for pvc to be bound.
+					o.Eventually(func() error {
+						pvc, err := oc.AdminKubeClient().CoreV1().PersistentVolumeClaims(oc.Namespace()).Get(context.Background(), "test-pvc", metav1.GetOptions{})
+						if err != nil {
+							return err
+						}
+						if pvc.Status.Phase != v1.ClaimBound {
+							return fmt.Errorf("PVC not bound")
+						}
+						return nil
+					})
+
+					for i := 0; i < t.successfulSnapshotsCreated; i++ {
+						err := createSnapshot(oc, oc.Namespace(), fmt.Sprintf("test-snapshot-%d", i), "test-pvc")
+						o.Expect(err).NotTo(o.HaveOccurred())
+					}
+
+					// Next snapshot creation should be over the set limit and fail.
+					err = createSnapshot(oc, oc.Namespace(), "test-snapshot-failed", "test-pvc")
+					o.Expect(err).NotTo(o.HaveOccurred())
+
+					readyToUse, err := oc.Run("get").Args("volumesnapshot/test-snapshot-failed", "-o", "jsonpath={.status.readyToUse}").Output()
+					o.Expect(err).NotTo(o.HaveOccurred())
+
+					errMsg, err := oc.Run("get").Args("volumesnapshot/test-snapshot-failed", "-o", "jsonpath={.status.error.message}").Output()
+					o.Expect(err).NotTo(o.HaveOccurred())
+
+					e2e.Logf("VolumeSnapshot error message: %s readyToUse %s", errMsg, readyToUse)
+					if !strings.Contains(errMsg, "failed to take snapshot of the volume") && readyToUse != "false" {
+						e2e.Failf("VolumeSnapshot \"test-snapshot-failed\" should have failed and should not be ready to use")
+					}
+				}
+			})
+
+		}
+	})
+})
+
+func setClusterCSIDriverSnapshotOptions(oc *exutil.CLI, snapshotOptions string, value int) {
+	patch := []byte(fmt.Sprintf("{\"spec\":{\"driverConfig\":{\"vSphere\":{\"%s\": %d}}}}", snapshotOptions, value))
+	_, err := oc.AdminOperatorClient().OperatorV1().ClusterCSIDrivers().Patch(context.Background(), providerName, types.MergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		e2e.Failf("failed to patch ClusterCSIDriver: %v", err)
+	}
+}
+
+func loadAndCheckCloudConf(oc *exutil.CLI, sectionName string, keyName string, expectedValue int) error {
+	cm, err := oc.AdminKubeClient().CoreV1().ConfigMaps("openshift-cluster-csi-drivers").Get(context.Background(), "vsphere-csi-config", metav1.GetOptions{})
+	if err != nil {
+		e2e.Failf("failed to get ConfigMap: %v", err)
+		return err
+	}
+
+	cloudConfData, ok := cm.Data["cloud.conf"]
+	if !ok {
+		return fmt.Errorf("cloud.conf key not found in ConfigMap")
+	}
+
+	cfg, err := ini.Load([]byte(cloudConfData))
+	if err != nil {
+		e2e.Failf("failed to load cloud.conf: %v", err)
+		return err
+	}
+
+	section, err := cfg.GetSection(sectionName)
+	if err != nil {
+		return fmt.Errorf("section %s not found in cloud.conf: %v", sectionName, err)
+	}
+
+	key, err := section.GetKey(keyName)
+	if err != nil {
+		return fmt.Errorf("key %s not found in section %s: %v", keyName, sectionName, err)
+	}
+
+	o.Expect(key.String()).To(o.Equal(fmt.Sprintf("%d", expectedValue)))
```

Comment:
You should check that the other keys are *not* set.
I wonder what `section.KeysHash()` returns. Is it the same map as `t.cloudConfigOptions`, if you used `map[string]string` instead of `int` as I suggest? Will `DeepEqual` work then?

Reply:
DeepEqual() seems to work well in this case, thanks for the suggestion.